keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+---
+title: Out-of-Tree Credential Providers
+authors:
+  - "@irozzo-1A"
+  - "@youssefazrak"
+owning-sig: sig-cloud-provider
+reviewers:
+  - "@cheftako"
+approvers: TBD
+editor: TBD
+creation-date: 2020-09-28
+last-updated: 2020-09-28
+status: draft
+---
+
+# Extending Apiserver Network Proxy to handle traffic originated from Node network
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Definitions](#definitions)
+- [Proposal](#proposal)
+  - [Traffic flow](#traffic-flow)
+  - [Agent additional flags](#agent-additional-flags)
+  - [Handling the Traffic from the Pods to the Agent](#handling-the-traffic-from-the-pods-to-the-agent)
+  - [Handling the Traffic from the Kubelet to the Agent](#handling-the-traffic-from-the-kubelet-to-the-agent)
+  - [Deployment Model](#deployment-model)
+  - [Listening Interface for Konnectivity agent](#listening-interface-for-konnectivity-agent)
+  - [Authentication](#authentication)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Allow list](#allow-list)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+The goal of this proposal is to allow traffic to flow from the Node Network to the Master Network.
+
+## Motivation
+
+API server network proxy has been originally introduced to allow running the cluster nodes on distinct isolated networks with respect to the one hosting the control plane components. This provides a way to handle traffic originating from the Kube API Server and going to the node networks. When using this setup, there are no other options than to directly expose the KAS to the Internet or setting up a VPN to handle traffic originated from the cluster nodes (i.e. Kubelet, pods). This could lead to security risks or complicated setups. 
```

Comment:
Regarding the "security risks": I think that exposing Konnectivity Proxy Server would add an additional layer of security, given that the channels with proxy agents are secured with mTLS or Token authentication.
The KAS is not directly exposed to the internet but only accessible from "secured" networks. This would protect it, for instance, from KAS misconfigurations or vulnerabilities that could expose sensitive information 
 and/or access to unauthenticated users.

IMHO on a higher level, if we believe that exposing the Kubelet to the internet brings security risks, the same should hold for the KAS.

Reply:
Regarding the "complicated setups": You are right we did not elaborate enough on this point, and "complicated" is not the appropriate term to be used here. We will amend the proposal.
The point we want to make is that right now there is no standard solution for this kind of setup. It is possible to rely on VPNs for example to achieve a similar goal, but this requires specific implementations. What we propose here is to build on top of what we have, and having a consistent approach for master to node and node to master communications.